stdlib/result.mli

```diff
@@ -0,0 +1,96 @@
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Result values.
+
+    Result values handle computation results and errors in an explicit
+    and declarative manner without resorting to exceptions.
+
+    @since 4.08 *)
+
+(** {1:results Results} *)
+
+type ('a, 'e) t = ('a, 'e) result = Ok of 'a | Error of 'e
+(** The type for result values. Either a value [Ok v] or an error [Error e]. *)
+
+val ok : 'a -> ('a, 'e) result
+(** [ok v] is [Ok v]. *)
+
+val error : 'e -> ('a, 'e) result
+(** [error e] is [Error e]. *)
+
+val value : ('a, 'e) result -> default:'a -> 'a
+(** [value r ~default] is [v] if [r] is [Ok v] and [default] otherwise. *)
+
+val get_ok : ('a, 'e) result -> 'a
+(** [get_ok r] is [v] if [r] is [Ok v] and @raise Invalid_argument
+    otherwise. *)
+
+val get_error : ('a, 'e) result -> 'e
+(** [get_error r] is [e] if [r] is [Error e] and @raise Invalid_argument
+    otherwise. *)
+
+val bind : ('a, 'e) result -> ('a -> ('b, 'e) result) -> ('b, 'e) result
+(** [bind r f] is [Ok (f v)] if [r] is [Ok v] and [r] if [r] is [Error _]. *)
+
+val join : (('a, 'e) result, 'e) result -> ('a, 'e) result
+(** [join rr] is [r] if [rr] is [Ok r] and [rr] if [rr] is [Error _]. *)
+
+val map : ('a -> 'b) -> ('a, 'e) result -> ('b, 'e) result
+(** [map f r] is [Ok (f v)] if [r] is [Ok v] and [r] if [r] is [Error _]. *)
+
+val fold : ('a -> 'c) -> ('e -> 'c) -> ('a, 'e) result -> 'c
+(** [fold ok error r] is [ok v] if [r] is [Ok v] and [error e] if [r]
+    is [Error e]. *)
+
+val iter : ('a -> unit) -> ('a, 'e) result -> unit
+(** [iter f r] is [f v] if [r] is [Ok v] and [()] otherwise. *)
+
+val recover : ('e -> 'a) -> ('a, 'e) result -> ('a, 'f) result
+(** [recover f r] is [r] if [r] is [Ok v] and [Ok (f e)] if [r] is
+    [Error e]. *)
+
+(** {1:preds Predicates and comparisons} *)
+
+val is_ok : ('a, 'e) result -> bool
+(** [is_ok r] is [true] iff [r] is [Ok _]. *)
+
+val is_error : ('a, 'e) result -> bool
+(** [is_error r] is [true] iff [r] is [Error _]. *)
+
+val equal :
+  ok:('a -> 'a -> bool) -> error:('e -> 'e -> bool) -> ('a, 'e) result ->
+  ('a, 'e) result -> bool
+(** [equal ~ok ~error r0 r1] tests equality of [r0] and [r1] using [ok]
+    and [error] to respectively compare values wrapped by [Ok _] and
+    [Error _]. *)
+
+val compare :
+  ok:('a -> 'a -> int) -> error:('e -> 'e -> int) -> ('a, 'e) result ->
+  ('a, 'e) result -> int
+(** [compare ~ok ~error r0 r1] totally orders [r0] and [r1] using [ok] and
+    [error] to respectively compare values wrapped by [Ok _ ] and [Error _].
+    [Ok _] values are smaller than [Error _] values. *)
+
+(** {1:convert Converting} *)
+
+val to_option : ('a, 'e) result -> 'a option
```

Comment:
This could also be called `get_ok_opt` to be in line with other Stdlib functions that come with both a raising and an option-returning variant.  But I'm not against `to_option`.  In that case, what about `of_option` (I know there is also Option.to_result, but one might expect to have both to_option/of_option at the same place)?

Reply:
I'm fine with people using `Option.to_result` for that. That way we just have `to_` functions in both `Option` and `Result`. If there is strong desire I can add both `Option.of_result` and `Result.of_option`.